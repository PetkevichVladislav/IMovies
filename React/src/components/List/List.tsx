import { Genre } from "../../models/enum/MenuGenre";
import { useMovieContext } from "../MovieListPage/MovieListPage";
import "./List.scss";

import { FC } from "react";

const selectedItemClassName = "list__item list__item--selected";
const unselectedItemClassName = "list__item";

interface IGenreListProps {
    genreNames: Genre[],
    selectedItemColor?: string,
}

export const List: FC<IGenreListProps> = ({ genreNames, selectedItemColor}) => {
    const movieContextModel = useMovieContext();

    const listGenres = genreNames.map((genre) => {
        const className = genre === movieContextModel?.selectedGenre ? selectedItemClassName : unselectedItemClassName;
        return <li key={genre} className={className} onClick={() => selectItem(genre)}>{genre}</li>;
    });

    const selectItem = (genreName: string) => {
        const selectedGenre = genreName as Genre;
        movieContextModel?.setSelectedGenre(selectedGenre);
    }

    const listClassNames : string[] = ["list"];
    if(selectedItemColor != null) {
        listClassNames.push(selectedItemColor);
    }
    return (
        <ul className={listClassNames.join(" ")}>
            {listGenres}
        </ul>
    );
}

export default List;