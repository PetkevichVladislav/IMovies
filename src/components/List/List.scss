.list {
    position: static;
    margin: 0px 0px 0px 0px;
    padding: 0px;
    display: flex;
    justify-content: space-between;

    &__item {
        list-style-type: none;
        margin: 20px 15px 20px 15px;
        color: #FFFFFF;
        position: relative;

        &--selected,
        &:hover {
            color: #F65261;

            &::after {
                content: "";
                position: absolute;
                bottom: -23px;
                width: 100%;
                height: 3px;
                left: 0px;
                background-color: #F65261;
            }
        }
    }
}

.list.pink {
    .list__item {

        &:hover {
            color: #F65261;

            &::after {
                background-color: #F65261;
            }
        }
    }


    &--selected,
    &:hover {
        color: #F65261;

        &::after {
            background-color: #F65261;
        }
    }
}

.list.blue {
    .list__item {

        &:hover {
            color: #52e3f6;

            &::after {
                background-color: #52e3f6;
            }
        }

        &--selected,
        &:hover {
            color: #52e3f6;

            &::after {
                background-color: #52e3f6;
            }
        }
    }
}