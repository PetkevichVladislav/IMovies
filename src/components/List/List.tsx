import "./List.scss";

import { FC, useState } from "react";

const selectedItemClassName = "list__item list__item--selected";
const unselectedItemClassName = "list__item";

interface IGenreListProps {
    itemNames: string[],
    preselectedItemName: string,
    onSelect: (itemName: string) => void,
}

const List: FC<IGenreListProps> = ({ itemNames, preselectedItemName, onSelect }) => {
    const [selectedItemName, setSelectedItemName] = useState<string>(preselectedItemName);
    const listItems = itemNames.map((itemName) => {
        const className = itemName === selectedItemName ? selectedItemClassName : unselectedItemClassName;
        return <li key={itemName} className={className} onClick={() => selectItem(itemName)}>{itemName}</li>;
    });

    const selectItem = (itemName: string) => {
        setSelectedItemName(itemName);
        onSelect(itemName);
    }

    return (
        <ul className="list">
            {listItems}
        </ul>
    );
}

export default List;