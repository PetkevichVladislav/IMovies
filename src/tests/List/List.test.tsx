import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import List from '../../components/List/List';

const selectedItemClassName = "list__item--selected";

describe("List", () => {
    const itemNames : string[] = ["Action", "Adventure", "Comedy", "Drama", "Sci-Fi", "Thriller", "Western"];
    const preselectedItemName = "Sci-Fi";
    const onSelectCallback = jest.fn();

    it("When data is valid component should renders correctly", () =>{
        const tree = renderer.create(<List itemNames={itemNames} preselectedItemName={preselectedItemName} onSelect={onSelectCallback}/>)
        .toJSON();
        expect(tree).toMatchSnapshot();;
    });

    it("when list doesn't contains passed elements preselected element should not be rendered", () =>{
        const itemNames : string[] = [];
        render(<List itemNames={itemNames} preselectedItemName={preselectedItemName} onSelect={onSelectCallback}/>);
        const list = screen.getByRole("list");
        expect(list).toBeEmptyDOMElement();
    });

    it("when list contains passed elements, the elements should be rendered", () =>{
        render(<List itemNames={itemNames} preselectedItemName={preselectedItemName} onSelect={onSelectCallback}/>);
        const listElements = screen.getAllByRole("listitem");
        expect(listElements.length === itemNames.length).toBeTruthy();
        expect(screen.getByText("Action")).toBeInTheDocument();
        expect(screen.getByText("Adventure")).toBeInTheDocument();
        expect(screen.getByText("Comedy")).toBeInTheDocument();
        expect(screen.getByText("Drama")).toBeInTheDocument();
        expect(screen.getByText("Sci-Fi")).toBeInTheDocument();
        expect(screen.getByText("Thriller")).toBeInTheDocument();
        expect(screen.getByText("Western")).toBeInTheDocument();
    });

    it("when list with elements passed, preselected element should be choosen", () =>{
        render(<List itemNames={itemNames} preselectedItemName={preselectedItemName} onSelect={onSelectCallback}/>);
        const selectedItem = screen.getByText(preselectedItemName);
        expect(selectedItem).toHaveClass(selectedItemClassName);
    });

    it("when list with elements, passed preselected element should be only one", () =>{
        const expectedCount : number = 1;
        render(<List itemNames={itemNames} preselectedItemName={preselectedItemName} onSelect={onSelectCallback}/>);
        const listItems = screen.getAllByRole("listitem").filter(item => item.classList.contains(selectedItemClassName));
        expect(listItems.length).toBe(expectedCount);
    });

    it("when clicks by element, callback function should return clicked element name", () =>{
        render(<List itemNames={itemNames} preselectedItemName={preselectedItemName} onSelect={onSelectCallback}/>);
        const itemToSelectName : string = 'Western';
        const itemToSelect = screen.getByText(itemToSelectName);
        fireEvent.click(itemToSelect);
        expect(onSelectCallback).toHaveBeenCalledWith(itemToSelectName);
    });

    it("when clicks by element, it should contains selected class", () =>{
        render(<List itemNames={itemNames} preselectedItemName={preselectedItemName} onSelect={onSelectCallback}/>);
        const itemToSelectName : string = 'Western';
        const itemToSelect = screen.getByText(itemToSelectName);
        fireEvent.click(itemToSelect);
        expect(itemToSelect).toHaveClass(selectedItemClassName);
    });
});