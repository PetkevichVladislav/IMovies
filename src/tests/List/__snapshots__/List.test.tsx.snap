// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List When data is valid component should renders correctly 1`] = `
<ul
  className="list"
>
  <li
    className="list__item"
    onClick={[Function]}
  >
    Action
  </li>
  <li
    className="list__item"
    onClick={[Function]}
  >
    Adventure
  </li>
  <li
    className="list__item"
    onClick={[Function]}
  >
    Comedy
  </li>
  <li
    className="list__item"
    onClick={[Function]}
  >
    Drama
  </li>
  <li
    className="list__item list__item--selected"
    onClick={[Function]}
  >
    Sci-Fi
  </li>
  <li
    className="list__item"
    onClick={[Function]}
  >
    Thriller
  </li>
  <li
    className="list__item"
    onClick={[Function]}
  >
    Western
  </li>
</ul>
`;
